### PR TITLE
[api-minor] Refactor fetching of built-in CMaps to utilize a factory on the `display` side instead, to allow users of the API to provide a custom CMap loading factory (e.g. for use with Node.js)

### LINKED ISSUE #4794: cmap files are loaded multiple times from the server
When opening ./test/pdfs/mao.pdf with the viewer, the cmap files are loaded multiple (8) times (even when the document contains only a single page).

I guess that this can be optimized.
screenshot-mao pdf - nightly-1

[PDF.js: 1.0.173, Linux mint]

### PR DESCRIPTION
**Pull request by Snuffleupagus**
Currently the built-in CMap files are loaded in `src/core/cmap.js` using `XMLHttpRequest` directly. For some environments that might be a problem, hence this patch refactors that to instead use a factory to load built-in CMaps on the main thread and message the data to the worker thread.

This is inspired by other recent work, e.g. the addition of the `CanvasFactory`, and to a large extent on the IRC discussion starting at http://logs.glob.uno/?c=mozilla%23pdfjs&s=12+Oct+2016&e=12+Oct+2016#c53010.

*Please note:* While it certainly may be possible to improve the patch, it does work just fine locally as-is, i.e. the viewer still works and all unit/font/reference tests pass.

@yurydelendik Since this PR attempts to implement your idea from http://logs.glob.uno/?c=mozilla%23pdfjs&s=12+Oct+2016&e=12+Oct+2016#c53018, I'd appreciate if you could provide feedback on the implementation when you've got time.

***Edit:*** Based on this PR, we should e.g. be able to get the CMap unit-tests running on Travis, by providing a custom Node.js factory for the unit-test.

***Edit 2:*** Also fixes #4794, courtesy of the second commit.

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/pdf.js/8064)

### CODE REVIEW

**Comment by Snuffleupagus** [comment]
Note that based on this patch, it would also be quite easy to add a simple cache to avoid having to load the *same* CMap files over and over, i.e. it could help with addressing issue #4794.

Although, I'm not sure if the intent of that issue was that we should somehow cache the *parsed* CMaps, or if it'd suffice if the raw CMaps files are cached. Anyway, something like the diff below would at least reduce the number of file loads considerable for some PDF files, without the need for any larger refactoring.
```diff
diff --git a/src/core/document.js b/src/core/document.js
index 9b2e4ea..8d482a3 100644
--- a/src/core/document.js
+++ b/src/core/document.js
@@ -71,13 +71,15 @@ var Page = (function PageClosure() {
   var DEFAULT_USER_UNIT = 1.0;
   var LETTER_SIZE_MEDIABOX = [0, 0, 612, 792];
 
-  function Page(pdfManager, xref, pageIndex, pageDict, ref, fontCache) {
+  function Page(pdfManager, xref, pageIndex, pageDict, ref, fontCache,
+                builtInCMapCache) {
     this.pdfManager = pdfManager;
     this.pageIndex = pageIndex;
     this.pageDict = pageDict;
     this.xref = xref;
     this.ref = ref;
     this.fontCache = fontCache;
+    this.builtInCMapCache = builtInCMapCache;
     this.evaluatorOptions = pdfManager.evaluatorOptions;
     this.resourcesPromise = null;
 
@@ -248,6 +250,7 @@ var Page = (function PageClosure() {
                                                   handler, this.pageIndex,
                                                   this.idFactory,
                                                   this.fontCache,
+                                                  this.builtInCMapCache,
                                                   this.evaluatorOptions);
 
       var dataPromises = Promise.all([contentStreamPromise, resourcesPromise]);
@@ -315,6 +318,7 @@ var Page = (function PageClosure() {
                                                     handler, self.pageIndex,
                                                     self.idFactory,
                                                     self.fontCache,
+                                                    self.builtInCMapCache,
                                                     self.evaluatorOptions);
 
         return partialEvaluator.getTextContent(contentStream,
@@ -551,9 +555,10 @@ var PDFDocument = (function PDFDocumentClosure() {
       this.xref.parse(recoveryMode);
       var self = this;
       var pageFactory = {
-        createPage: function (pageIndex, dict, ref, fontCache) {
+        createPage: function (pageIndex, dict, ref, fontCache,
+                              builtInCMapCache) {
           return new Page(self.pdfManager, self.xref, pageIndex, dict, ref,
-                          fontCache);
+                          fontCache, builtInCMapCache);
         }
       };
       this.catalog = new Catalog(this.pdfManager, this.xref, pageFactory);
diff --git a/src/core/evaluator.js b/src/core/evaluator.js
index ca26861..9078f11 100644
--- a/src/core/evaluator.js
+++ b/src/core/evaluator.js
@@ -169,18 +169,29 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
   };
 
   function PartialEvaluator(pdfManager, xref, handler, pageIndex,
-                            idFactory, fontCache, options) {
+                            idFactory, fontCache, builtInCMapCache, options) {
     this.pdfManager = pdfManager;
     this.xref = xref;
     this.handler = handler;
     this.pageIndex = pageIndex;
     this.idFactory = idFactory;
     this.fontCache = fontCache;
+    this.builtInCMapCache = builtInCMapCache;
     this.options = options || DefaultPartialEvaluatorOptions;
 
     this.fetchBuiltInCMap = function (name) {
+      var cachedCMap = builtInCMapCache[name];
+      if (cachedCMap) {
+        return Promise.resolve(cachedCMap);
+      }
       return handler.sendWithPromise('FetchBuiltInCMap', {
         name: name,
+      }).then(function (data) {
+        if (data.isBinary) {
+          // Given the size of uncompressed CMaps, we only cached binary CMaps.
+          builtInCMapCache[name] = data;
+        }
+        return data;
       });
     };
   }
diff --git a/src/core/obj.js b/src/core/obj.js
index c45e543..74f404b 100644
--- a/src/core/obj.js
+++ b/src/core/obj.js
@@ -71,8 +71,8 @@ var Catalog = (function CatalogClosure() {
     this.xref = xref;
     this.catDict = xref.getCatalogObj();
     this.fontCache = new RefSetCache();
-    assert(isDict(this.catDict),
-      'catalog object is not a dictionary');
+    this.builtInCMapCache = Object.create(null);
+    assert(isDict(this.catDict), 'catalog object is not a dictionary');
 
     // TODO refactor to move getPage() to the PDFDocument.
     this.pageFactory = pageFactory;
@@ -418,6 +418,8 @@ var Catalog = (function CatalogClosure() {
     },
 
     cleanup: function Catalog_cleanup() {
+      this.builtInCMapCache = Object.create(null);
+
       var promises = [];
       this.fontCache.forEach(function (promise) {
         promises.push(promise);
@@ -438,7 +440,8 @@ var Catalog = (function CatalogClosure() {
             var dict = a[0];
             var ref = a[1];
             return this.pageFactory.createPage(pageIndex, dict, ref,
-                                               this.fontCache);
+                                               this.fontCache,
+                                               this.builtInCMapCache);
           }.bind(this)
         );
       }
diff --git a/test/unit/document_spec.js b/test/unit/document_spec.js
index cdec2ae..11f3fba 100644
--- a/test/unit/document_spec.js
+++ b/test/unit/document_spec.js
@@ -34,11 +34,13 @@ describe('document', function () {
       var page1 = new Page(/* pdfManager = */ { }, /* xref = */ null,
                            /* pageIndex = */ 0,
                            /* pageDict = */ null, /* ref = */ null,
-                           /* fontCache = */ null);
+                           /* fontCache = */ null,
+                           /* builtInCMapCache = */ null);
       var page2 = new Page(/* pdfManager = */ { }, /* xref = */ null,
                            /* pageIndex = */ 1,
                            /* pageDict = */ null, /* ref = */ null,
-                           /* fontCache = */ null);
+                           /* fontCache = */ null,
+                           /* builtInCMapCache = */ null);
 
       var idFactory1 = page1.idFactory, idFactory2 = page2.idFactory;
```

File: src/display/dom_utils.js [hunk 1]
```diff
@@ -66,6 +66,59 @@ DOMCanvasFactory.prototype = {
...
+          request.responseType = 'arraybuffer';
+        }
+        request.onreadystatechange = function () {
+          if (request.readyState === XMLHttpRequest.DONE &&
+              (request.status === 200 || request.status === 0)) {
+            var data;
+            if (this.binary && request.response) {
+              data = new Uint8Array(request.response);
+            } else if (!this.binary && request.responseText) {
+              var arr = Array.prototype.map.call(request.responseText,
```
Comment (yurydelendik) [inline src/display/dom_utils.js]: You can use `stringToBytes` here.

**Review by yurydelendik — APPROVED** [review]
Looks good with the stringToBytes change.

**Comment by yurydelendik** [comment]
I don't think anything changes here, but we also need to be sure permission restrictions stay the same when worker is loaded from foreign origin.

**Comment by yurydelendik** [comment]
For future, we can opened an issue to create a separate factory and set of js-files for each map to be loaded cross-origin (e.g. these js files will have binary encoded array which will be decoded when loaded via script tag, the files can be generated during pdfjs-dist build and use any PDF compression, e.g. have it as ASCII85 and DEFLATE'd; maybe we shall change bool 'isBinary' field to enum)

File: src/display/dom_utils.js (hunk 1 above)
Comment (Snuffleupagus) [inline src/display/dom_utils.js]: Good point, fixed now.

**Comment by Snuffleupagus** [comment]
> [...] maybe we shall change bool 'isBinary' field to enum

Good idea, I've attempted to implement this, an interdiff should be available through https://reviewable.io/reviews/mozilla/pdf.js/8064.
Thank you for the review, and please let me know if this is what you had in mind!

**Comment by Snuffleupagus** [comment]
@yurydelendik Even though you already approved this PR, would you mind quickly looking over it again before landing it?
Please note that based on your review comments, I've made a couple of changes as outlined in https://github.com/mozilla/pdf.js/pull/8064#issuecomment-279782921

**Comment by yurydelendik** [comment]
> @yurydelendik Even though you already approved this PR, would you mind quickly looking over it again before we land it?

lgtm

**Comment by yurydelendik** [comment]
Thank you for the patch.

**Comment by timvandermeij** [comment]
Nice, this has made a massive difference for me with loading `test/pdfs/mao.pdf`. The loading went down from over 4 seconds to approximately 1 second.

**Pull request merged**